src/coreclr/System.Private.CoreLib/src/System/Diagnostics/StackFrameHelper.cs

```diff
@@ -112,17 +144,12 @@ internal void InitializeSourceInfo(bool fNeedFileInfo, Exception? exception)
                     Interlocked.CompareExchange(ref s_stackTraceSymbolsCache, CreateStackTraceSymbols(), null);
                 }
 
-                for (int index = 0; index < iFrameCount; index++)
-                {
-                    // If there was some reason not to try get the symbols from the portable PDB reader like the module was
-                    // ENC or the source/line info was already retrieved, the method token is 0.
-                    if (rgiMethodToken![index] != 0)
-                    {
-                        GetSourceLineInfo(s_stackTraceSymbolsCache!, rgAssembly![index], rgAssemblyPath![index]!, rgLoadedPeAddress![index], rgiLoadedPeSize![index], rgiIsFileLayout![index],
-                            rgInMemoryPdbAddress![index], rgiInMemoryPdbSize![index], rgiMethodToken![index],
-                            rgiILOffset![index], out rgFilename![index], out rgiLineNumber![index], out rgiColumnNumber![index]);
-                    }
-                }
+                GetSourceLineInfo(s_stackTraceSymbolsCache!, rgAssembly![index], rgAssemblyPath![index]!, rgLoadedPeAddress![index], rgiLoadedPeSize![index], rgiIsFileLayout![index],
+                    rgInMemoryPdbAddress![index], rgiInMemoryPdbSize![index], rgiMethodToken![index],
+                    rgiILOffset![index], out rgFilename![index], out rgiLineNumber![index], out rgiColumnNumber![index]);
+
+                // Make sure we mark down that debug information for this frame was resolved
+                rgiMethodToken[index] = 0;
```

Comment:
Does this need to be thread safe? Other parts of the change use Interlocked.CompareExchange that suggests this is expected to be thread safe.

Reply:
Most likely. I think it should be sufficient to outline the `out` parameters to be assigned after the call returns, and make the read/write to `rgiMethodToken[index]` volatile? The data race will only see multiple threads executing here, and both *should* result in the same values, so I don't think anything more is needed.

Reply:
Should be done now